### DISCUSSION TITLE
OutParameterUndefinedWhenAttribute(bool value)

### DISCUSSION THREAD

**Discussion by TactiTac0z** [dotnet/csharplang#8530] [Language Ideas]
It's pretty common in C# that we use / write methods similar to `bool TryGetValue(out T value)`

Like for dictionaries, if the return value is false the value gets assigned to default or null.
Depending on inlineability of a method such as that, the runtime may or may not inline the default assignment away atleast if a local variable is used as the parameter.

It would be quite convenient even just at a language level, if we could enforce the consumer via static analysis that an argument supplied to such a method is only definitively assigned if the method return true (or false), but also for a little bit of performance improvements.
MemberNotNullWhen technically does this at a language level, although that wouldn't work for value types where a performance improvement would actually matter.

```cs
bool TryGetValue([OutParameterUndefinedWhen(false)] out (string, string) result)
{
  // Illegal, error CSxxxx: out parameter must be assigned before control leaves the method (unless the return value is 'false')
  return true;
  // .....................................

  if (world.exist)
  {
    result = ("Hallo", "World");
    return true;
  }
  else
    // allowed
    return false;
}

void Main()
{
  TryGetValue(out var result);
  // Illegal, error csxxxx: variable 'result' may be undefined
  Console.WriteLine(result);

  // allowed
  if (TryGetValue(out var result))
    Console.WriteLine(result);
}
```
### Draw backs
To be fair there is a quite a bit of draw backs here.
I admit this might be better suited using analyzers and unsafe code when I really want those few clock cycles.
I'm gonna post it anyway in case it can be used for something anyway.
These issues could theoretically be fixed with actual runtime support.

Languages and version that don't understand this new attribute might try to dereference an uninitialized memory block, although to be fair initLocals is emitted by default in pretty much all compilers but still.

If an interior pointer is provided as the argument an actual observable behavior is caused if a consumer code depend on that field or array being set to default. So it would likely not be safe to change bcl methods to use this.

**Comment by FaustVX** [comment]
You can use [`Unsafe.SkipInit()`](https://learn.microsoft.com/dotnet/api/system.runtime.compilerservices.unsafe.skipinit) and the `out` parameter will be untouched as you can see in [sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQAwAJMCMAdAEoCuMCUYApkQMID2YADlADY0BOAytwG5QAxjRQBuALAAoaZgDM+AEy560gN7TcW/AsIA2fABZcAWQAUjMglyxrUAJSbtzgLwA+XAFUYKCADM6HgBrKBYASRgoBAsrG3tJKQBfIA==)

**Reply by colejohnson66** [reply]
As the name implies, this is ***unsafe***. In this case, the calling function will have zero-inited the area the `out` variable points, so it's *probably* safe. But the overhead of properly initing it is almost certainly negligible; Just `outVar = default` at the start of your method.
